### PR TITLE
Fix inner query sort limit

### LINKED ISSUE #16921: Aggregate does not yield last date in Insights
### Describe the Bug

Since the upgrade to 9.22, one of my insights using a metric panel does not aggregate a date field to show the "last" date. Instead it shows the first date (may be the default behaviour).

### Background/Context
I have two tables (Collections)
1. Notes
2. Procedures

Whenever I need to log anything for my pet (vet visits, medications, procedures, etc.), I create a note. Hence the two tables are connected together in a many-to-many relationship, using a "Notes" table field which is also called "Procedures" (which relates to a "Procedures" table field also called "Notes").

In this example, I have administered a deworming tablet (so the procedure is called "Internal Anti-parasitic immunization").

### To Reproduce

I have a "metric" panel in insights to show me the date when the last "internal anti-parasitic immunization" was administered. I have done this by creating a "metric" panel, which looks in the "Notes" table, for the "Date" field, which is aggregated using the "last" function. I have also sorted by "Date" (but I feel this is irrelevant). I have added the filter `Procedures` → `Procedures ID` → `Procedure Name` `Contains` `internal`.

### Errors Shown

None.

### What version of Directus are you using?

9.22.0 & 9.22.1

### What version of Node.js are you using?

Included in Docker

### What database are you using?

MariaDB 10.3.36

### What browser are you using?

Firefox 108

### How are you deploying Directus?

Docker

### Note

I have rolled back to 9.21.2, and the behavious is as expected (shows the last date correctly).

### LINKED ISSUE #16991: Order in nested filters with pagination
### Describe the Bug

After updating to version 9.22.1 when using a filter with nested fields, and using pagination, the ordering is wrong.

I believe this bug is related to this PR: https://github.com/directus/directus/pull/14798

### To Reproduce

Use this branch: https://github.com/diegoleme/directus-reproduction/tree/bug-9.22.1

Up the directus with docker compose
```sh
docker-cpmpose up
```

Make a request sorting by date created
```sh
curl --location -g --request GET 'http://localhost:8055/items/order?sort=date_created&limit=2&filter={"delivery_point":{"contact":{"_eq":"c989b26a-41f9-4f25-b304-f5dd27323dbd"}}}&fields=id,date_created,name'
```

Received response:
```json
{
    "data": [
        {
            "id": "186941f7-59cd-43de-accc-cf50f4c07787",
            "date_created": "2022-12-30T16:24:33.319Z",
            "name": "Order 2"
        },
        {
            "id": "20d1e71b-0f92-4515-84d7-edfe608c4a04",
            "date_created": "2022-12-30T16:25:28.690Z",
            "name": "Order 6"
        }
    ]
}
```

Expected response:
```json
{
    "data": [
        {
            "id": "f3aa21ab-b3eb-4de3-978e-990131cf8c2c",
            "date_created": "2022-12-30T16:23:24.876Z",
            "name": "Order 1"
        },
        {
            "id": "186941f7-59cd-43de-accc-cf50f4c07787",
            "date_created": "2022-12-30T16:24:33.319Z",
            "name": "Order 2"
        }
    ]
}
```

In the executed query, the ordering is separated from the limit, and this causes the bug.
```sql
select `order`.`id`,
`order`.`date_created`,
`order`.`name`
from `order`
  inner join (
    select distinct `order`.`id`,
      `order`.`date_created` as `sort_lmopk`
    from `order`
      left join `delivery_points` as `qlier` on `order`.`delivery_point` = `qlier`.`id`
      left join `directus_users` as `iujfz` on `qlier`.`id` = `iujfz`.`delivery_point_contact`
    where (`iujfz`.`id` = ?)
    limit ?
  ) as `inner` on `order`.`id` = `inner`.`id`
order by `inner`.`sort_lmopk` asc [c989b26a-41f9-4f25-b304-f5dd27323dbd, 2]
```

### Errors Shown

_No response_

### What version of Directus are you using?

9.22.1

### What version of Node.js are you using?

18

### What database are you using?

SQLite

### What browser are you using?

Chrome

### How are you deploying Directus?

Docker

### PR DESCRIPTION
**Pull request by licitdev**
## Description

- [x] Fix sorting with limits
- [x] Fix MSSQL support
- [x] Fix M2A deep fetching of function fields in REST
- [x] Add & fix tests

Fixes #16921 
Fixes ENG-347
Fixes #16991
Fixes ENG-376

## Type of Change

- [x] Bugfix
- [ ] Improvement
- [ ] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [x] New / updated tests are included
- [x] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated. PR:

### CODE REVIEW

File: api/src/database/run-ast.ts [hunk 1]
```diff
@@ -297,37 +304,25 @@ async function getDBQuery(
 							originalCollectionName: getCollectionFromAlias(alias, aliasMap),
 						})
 					);
+
+					orderByString += `?? ${sortRecord.order}`;
+					orderByFields.push(
+						getColumn(knex, alias, field, false, schema, {
+							originalCollectionName: getCollectionFromAlias(alias, aliasMap),
+						})
+					);
```
[the remark this answers is not in the document]
Comment (licitdev) [reply api/src/database/run-ast.ts]: Yes we should! Thanks! 👍 Added in https://github.com/directus/directus/pull/16922/commits/58531ee1ed1623916430a970f0eab8ffd0505c2d

File: api/src/database/helpers/schema/dialects/mssql.ts [hunk 2]
```diff
@@ -2,6 +2,14 @@ import { Knex } from 'knex';
 import { SchemaHelper } from '../types';
 
 export class SchemaHelperMSSQL extends SchemaHelper {
+	applyLimit(rootQuery: Knex.QueryBuilder, limit: number): void {
+		if (limit === -1) {
+			rootQuery.limit(Number.MAX_SAFE_INTEGER);
+		} else {
+			rootQuery.limit(limit);
+		}
+	}
```
[the remark this answers is not in the document]
Comment (licitdev) [reply api/src/database/helpers/schema/dialects/mssql.ts]: That's very true. Comments can be very helpful in the helpers context. 😄 
Added a comment detailing the error that it overcomes in https://github.com/directus/directus/pull/16922/commits/7b8353cb3d9b94022de733c2ddee933894b15f40

File: api/src/database/helpers/schema/dialects/mysql.ts [hunk 3]
```diff
@@ -12,7 +12,7 @@ export class SchemaHelperMySQL extends SchemaHelper {
 		orderByFields: Knex.Raw[]
 	): Knex.QueryBuilder {
 		if (getDatabaseVersion()?.startsWith('5.7')) {
-			dbQuery.orderByRaw(`?? asc, ${orderByString.slice(9)}`, [`${table}.${primaryKey}`, ...orderByFields]);
+			dbQuery.orderByRaw(`?? asc, ${orderByString}`, [`${table}.${primaryKey}`, ...orderByFields]);
```
[the remark this answers is not in the document]
Comment (licitdev) [reply api/src/database/helpers/schema/dialects/mysql.ts:15]: Yup, this was the reason behind the string spacing improvement. Thanks for the heads up! 🙌

File: api/src/database/helpers/schema/dialects/mssql.ts [hunk 4]
```diff
@@ -2,6 +2,16 @@ import { Knex } from 'knex';
 import { SchemaHelper } from '../types';
 
 export class SchemaHelperMSSQL extends SchemaHelper {
+	applyLimit(rootQuery: Knex.QueryBuilder, limit: number): void {
+		// The ORDER BY clause is invalid in views, inline functions, derived tables, subqueries,
+		// and common table expressions, unless TOP, OFFSET or FOR XML is also specified.
+		if (limit === -1) {
+			rootQuery.limit(Number.MAX_SAFE_INTEGER);
```
Comment (br41nslug) [inline api/src/database/helpers/schema/dialects/mssql.ts:9]: Do we need to set a limit here? I believe a SQL query without explicit limit should return all items right? or am i missing a MS SQL edge case 🤔

Comment (licitdev) [reply api/src/database/helpers/schema/dialects/mssql.ts:9]: Yes we do, in order to allow for sorting in a subquery.
Technically we should be using `TOP (100) percent`, but it's not supported when using limits from Knex. So the MAX_SAFE_INTEGER is used.

Refer to the two comment lines prior.
https://github.com/directus/directus/blob/7b8353cb3d9b94022de733c2ddee933894b15f40/api/src/database/helpers/schema/dialects/mssql.ts#L6-L7

**Review by br41nslug — APPROVED** [review]
Tested for [ENG-376](https://linear.app/directus/issue/ENG-376) in MSSQL and SQLite.

**Review by azrikahar — APPROVED** [review]

**Pull request merged**